### PR TITLE
Do not use EphemeralKey on MacOS for loading certs

### LINKED ISSUE #1976: Unable to read Certificate from KeyVault when running in MacOS (M2 chip)
### Microsoft.Identity.Web Library

Microsoft.Identity.Web

### Microsoft.Identity.Web version

1.25.1

### Web app

Not Applicable

### Web API

Protected web APIs call downstream web APIs

### Token cache serialization

Distributed caches

### Description

I have a web api secured by AAD that calls Microsoft Graph API (downstream API).

```sharp
services.AddMicrosoftIdentityWebApiAuthentication(configuration)
                .EnableTokenAcquisitionToCallDownstreamApi()
                .AddMicrosoftGraph()
                .AddDistributedTokenCaches();
```

To call Graph, I'm using a Certificate in Azure KeyVault. The certificate is configured in settings.json:

```xml
    "ClientCertificates": [
      {
        "SourceType": "KeyVault",
        "KeyVaultUrl": "https://my-kv.vault.azure.net/",
        "KeyVaultCertificateName": "MyCertificate"
      }
    ]
```

Everything is working perfectly fine when running in Windows, and actually, the API is already in production, deployed to Azure. However, I'm trying to run it locally in a new __MacBook Air M2__ (Ventura 13.0.1) and I get the following error when _Microsoft.Identity.Web_ is trying to read the certificate from Azure KeyVault:

> System.PlatformNotSupportedException: This platform does not support loading with EphemeralKeySet. Remove the flag to allow keys to be temporarily created on disk.
         at Internal.Cryptography.Pal.AppleCertificatePal.FromBlob(ReadOnlySpan`1 rawData, SafePasswordHandle password, X509KeyStorageFlags keyStorageFlags)
         at System.Security.Cryptography.X509Certificates.X509Certificate..ctor(Byte[] rawData, String password, X509KeyStorageFlags keyStorageFlags)
         at System.Security.Cryptography.X509Certificates.X509Certificate2..ctor(Byte[] rawData, String password, X509KeyStorageFlags keyStorageFlags)
         at Microsoft.Identity.Web.DefaultCertificateLoader.LoadFromBase64Encoded(String certificateBase64, X509KeyStorageFlags x509KeyStorageFlags)
         at Microsoft.Identity.Web.DefaultCertificateLoader.LoadFromKeyVault(String keyVaultUrl, String certificateName, X509KeyStorageFlags x509KeyStorageFlags)
         at Microsoft.Identity.Web.DefaultCertificateLoader.LoadIfNeeded(CertificateDescription certificateDescription)
         at Microsoft.Identity.Web.DefaultCertificateLoader.<>c__DisplayClass5_0.<LoadFirstCertificate>b__0(CertificateDescription c)
         at System.Linq.Enumerable.TryGetFirst[TSource](IEnumerable`1 source, Func`2 predicate, Boolean& found)
         at System.Linq.Enumerable.FirstOrDefault[TSource](IEnumerable`1 source, Func`2 predicate)
         at Microsoft.Identity.Web.DefaultCertificateLoader.LoadFirstCertificate(IEnumerable`1 certificateDescriptions)
         at Microsoft.Identity.Web.TokenAcquisition.BuildConfidentialClientApplication(MergedOptions mergedOptions)
         at Microsoft.Identity.Web.TokenAcquisition.GetOrBuildConfidentialClientApplication(MergedOptions mergedOptions)
         at Microsoft.Identity.Web.TokenAcquisition.GetAuthenticationResultForUserAsync(IEnumerable`1 scopes, String authenticationScheme, String tenantId, String userFlow, ClaimsPrincipal user, TokenAcquisitionOptions tokenAcquisitionOptions)
         at Microsoft.Identity.Web.TokenAcquisitionAuthenticationProvider.AuthenticateRequestAsync(HttpRequestMessage request)
         at Microsoft.Graph.AuthenticationHandler.SendAsync(HttpRequestMessage httpRequestMessage, CancellationToken cancellationToken)
         at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
         at Microsoft.Graph.HttpProvider.SendRequestAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
         --- End of inner exception stack trace ---
         at Microsoft.Graph.HttpProvider.SendRequestAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
         at Microsoft.Graph.HttpProvider.SendAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
         at Microsoft.Graph.BaseRequest.SendRequestAsync(Object serializableObject, CancellationToken cancellationToken, HttpCompletionOption completionOption)
         at Microsoft.Graph.BaseRequest.SendAsync[T](Object serializableObject, CancellationToken cancellationToken, HttpCompletionOption completionOption)
         at Microsoft.Graph.UserTransitiveMemberOfCollectionWithReferencesRequest.GetAsync(CancellationToken cancellationToken)
         at Atlas.Connex.Infrastructure.Repositories.GraphUserRepository.IsCurrentUserTemplatesAdmin() in /Users/luisman/dev/atlas/src/Infrastructure/Connex.Infrastructure/Repositories/GraphUserRepository.cs:line 82
         at Atlas.Connex.Application.Templates.Queries.GetAllTemplates.GetAllTemplatesQueryHandler.Handle(GetAllTemplatesQuery request, CancellationToken cancellationToken) in /Users/luisman/dev/atlas/src/Core/Connex.Application/Templates/Queries/GetAllTemplates/GetAllTemplatesQuery.cs:line 47
         at Atlas.Connex.Application.Common.Behaviour.LoggingBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next) in /Users/luisman/dev/atlas/src/Core/Connex.Application/Common/Behaviour/LoggingBehaviour.cs:line 43
         at MediatR.Pipeline.RequestExceptionProcessorBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
         at MediatR.Pipeline.RequestExceptionProcessorBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
         at MediatR.Pipeline.RequestExceptionActionProcessorBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
         at MediatR.Pipeline.RequestExceptionActionProcessorBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
         at MediatR.Pipeline.RequestPostProcessorBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
         at MediatR.Pipeline.RequestPreProcessorBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
         at Atlas.Connex.Api.Controllers.TemplatesController.GetAllTemplates(AllTemplateRequest body) in /Users/luisman/dev/atlas/src/Api/Connex.Api/Controllers/TemplatesController.cs:line 316
         at lambda_method81(Closure , Object )
         at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.AwaitableObjectResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Logged|12_1(ControllerActionInvoker invoker)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()
      --- End of stack trace from previous location ---
         at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeFilterPipelineAsync()
      --- End of stack trace from previous location ---
         at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
         at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
         at Microsoft.AspNetCore.Authorization.Policy.AuthorizationMiddlewareResultHandler.HandleAsync(RequestDelegate next, HttpContext context, AuthorizationPolicy policy, PolicyAuthorizationResult authorizeResult)
         at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
         at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
         at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.<Invoke>g__Awaited|6_0(ExceptionHandlerMiddleware middleware, HttpContext context, Task task)

### Reproduction steps

1. Secure a WebAPI
2. Configure to call MS Graph API as downstream API
3. Configure an Azure KeyVault with certificate
4. Setup API to read the certificate from Az KeyVault

### Error message

System.PlatformNotSupportedException: This platform does not support loading with EphemeralKeySet. Remove the flag to allow keys to be temporarily created on disk.
         at Internal.Cryptography.Pal.AppleCertificatePal.FromBlob(ReadOnlySpan`1 rawData, SafePasswordHandle password, X509KeyStorageFlags keyStorageFlags)
         at System.Security.Cryptography.X509Certificates.X509Certificate..ctor(Byte[] rawData, String password, X509KeyStorageFlags keyStorageFlags)
         at System.Security.Cryptography.X509Certificates.X509Certificate2..ctor(Byte[] rawData, String password, X509KeyStorageFlags keyStorageFlags)
         at Microsoft.Identity.Web.DefaultCertificateLoader.LoadFromBase64Encoded(String certificateBase64, X509KeyStorageFlags x509KeyStorageFlags)
         at Microsoft.Identity.Web.DefaultCertificateLoader.LoadFromKeyVault(String keyVaultUrl, String certificateName, X509KeyStorageFlags x509KeyStorageFlags)
         at Microsoft.Identity.Web.DefaultCertificateLoader.LoadIfNeeded(CertificateDescription certificateDescription)
         at Microsoft.Identity.Web.DefaultCertificateLoader.<>c__DisplayClass5_0.<LoadFirstCertificate>b__0(CertificateDescription c)
         at System.Linq.Enumerable.TryGetFirst[TSource](IEnumerable`1 source, Func`2 predicate, Boolean& found)
         at System.Linq.Enumerable.FirstOrDefault[TSource](IEnumerable`1 source, Func`2 predicate)
         at Microsoft.Identity.Web.DefaultCertificateLoader.LoadFirstCertificate(IEnumerable`1 certificateDescriptions)
         at Microsoft.Identity.Web.TokenAcquisition.BuildConfidentialClientApplication(MergedOptions mergedOptions)
         at Microsoft.Identity.Web.TokenAcquisition.GetOrBuildConfidentialClientApplication(MergedOptions mergedOptions)
         at Microsoft.Identity.Web.TokenAcquisition.GetAuthenticationResultForUserAsync(IEnumerable`1 scopes, String authenticationScheme, String tenantId, String userFlow, ClaimsPrincipal user, TokenAcquisitionOptions tokenAcquisitionOptions)
         at Microsoft.Identity.Web.TokenAcquisitionAuthenticationProvider.AuthenticateRequestAsync(HttpRequestMessage request)
         at Microsoft.Graph.AuthenticationHandler.SendAsync(HttpRequestMessage httpRequestMessage, CancellationToken cancellationToken)
         at System.Net.Http.HttpClient.<SendAsync>g__Core|83_0(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationTokenSource cts, Boolean disposeCts, CancellationTokenSource pendingRequestsCts, CancellationToken originalCancellationToken)
         at Microsoft.Graph.HttpProvider.SendRequestAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
         --- End of inner exception stack trace ---
         at Microsoft.Graph.HttpProvider.SendRequestAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
         at Microsoft.Graph.HttpProvider.SendAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
         at Microsoft.Graph.BaseRequest.SendRequestAsync(Object serializableObject, CancellationToken cancellationToken, HttpCompletionOption completionOption)
         at Microsoft.Graph.BaseRequest.SendAsync[T](Object serializableObject, CancellationToken cancellationToken, HttpCompletionOption completionOption)
         at Microsoft.Graph.UserTransitiveMemberOfCollectionWithReferencesRequest.GetAsync(CancellationToken cancellationToken)
         at Atlas.Connex.Infrastructure.Repositories.GraphUserRepository.IsCurrentUserTemplatesAdmin() in /Users/luisman/dev/atlas/src/Infrastructure/Connex.Infrastructure/Repositories/GraphUserRepository.cs:line 82
         at Atlas.Connex.Application.Templates.Queries.GetAllTemplates.GetAllTemplatesQueryHandler.Handle(GetAllTemplatesQuery request, CancellationToken cancellationToken) in /Users/luisman/dev/atlas/src/Core/Connex.Application/Templates/Queries/GetAllTemplates/GetAllTemplatesQuery.cs:line 47
         at Atlas.Connex.Application.Common.Behaviour.LoggingBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next) in /Users/luisman/dev/atlas/src/Core/Connex.Application/Common/Behaviour/LoggingBehaviour.cs:line 43
         at MediatR.Pipeline.RequestExceptionProcessorBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
         at MediatR.Pipeline.RequestExceptionProcessorBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
         at MediatR.Pipeline.RequestExceptionActionProcessorBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
         at MediatR.Pipeline.RequestExceptionActionProcessorBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
         at MediatR.Pipeline.RequestPostProcessorBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
         at MediatR.Pipeline.RequestPreProcessorBehavior`2.Handle(TRequest request, CancellationToken cancellationToken, RequestHandlerDelegate`1 next)
         at Atlas.Connex.Api.Controllers.TemplatesController.GetAllTemplates(AllTemplateRequest body) in /Users/luisman/dev/atlas/src/Api/Connex.Api/Controllers/TemplatesController.cs:line 316
         at lambda_method81(Closure , Object )
         at Microsoft.AspNetCore.Mvc.Infrastructure.ActionMethodExecutor.AwaitableObjectResultExecutor.Execute(IActionResultTypeMapper mapper, ObjectMethodExecutor executor, Object controller, Object[] arguments)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeActionMethodAsync>g__Logged|12_1(ControllerActionInvoker invoker)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.<InvokeNextActionFilterAsync>g__Awaited|10_0(ControllerActionInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Rethrow(ActionExecutedContextSealed context)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ControllerActionInvoker.InvokeInnerFilterAsync()
      --- End of stack trace from previous location ---
         at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeNextResourceFilter>g__Awaited|25_0(ResourceInvoker invoker, Task lastTask, State next, Scope scope, Object state, Boolean isCompleted)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Rethrow(ResourceExecutedContextSealed context)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.Next(State& next, Scope& scope, Object& state, Boolean& isCompleted)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.InvokeFilterPipelineAsync()
      --- End of stack trace from previous location ---
         at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
         at Microsoft.AspNetCore.Mvc.Infrastructure.ResourceInvoker.<InvokeAsync>g__Logged|17_1(ResourceInvoker invoker)
         at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
         at Microsoft.AspNetCore.Authorization.Policy.AuthorizationMiddlewareResultHandler.HandleAsync(RequestDelegate next, HttpContext context, AuthorizationPolicy policy, PolicyAuthorizationResult authorizeResult)
         at Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
         at Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
         at Microsoft.AspNetCore.Diagnostics.ExceptionHandlerMiddleware.<Invoke>g__Awaited|6_0(ExceptionHandlerMiddleware middleware, HttpContext context, Task task)

### Id Web logs

_No response_

### Relevant code snippets

```csharp
Already included in Bug description
```

### Regression

_No response_

### Expected behavior

Certificate is loaded from KeyVault

### PR DESCRIPTION
**Pull request by bgavrilMS**
# do not use EphemeralKey on MacOS for loading certs

Summary of the changes (Less than 80 chars)

- do not use EphemeralKey on MacOS for loading certs. Instead, use the user key. 

## Description

{Detail}

Fixes #1976

### CODE REVIEW

File: src/Microsoft.Identity.Web.Certificate/CertificateLoaderHelper.cs [hunk 1]
```diff
@@ -6,29 +6,46 @@
...
-                x509KeyStorageFlags = X509KeyStorageFlags.MachineKeySet;
+            return X509KeyStorageFlags.MachineKeySet;
 #else
-                x509KeyStorageFlags = X509KeyStorageFlags.MachineKeySet | X509KeyStorageFlags.EphemeralKeySet;
-#endif
+            // This is for app developers using a Mac. MacOS does not support the EphemeralKeySet flag.
+            // See https://learn.microsoft.com/en-us/dotnet/standard/security/cross-platform-cryptography#write-a-pkcs12pfx
+            if (OsHelper.IsMacPlatform())
+            {
+                return X509KeyStorageFlags.DefaultKeySet;
```
Comment (bgavrilMS) [inline src/Microsoft.Identity.Web.Certificate/CertificateLoaderHelper.cs:44]: The default should be the user key set, which seems appropriate for a dev machine.

File: src/Microsoft.Identity.Web.Diagnostics/OsHelper.cs [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Runtime.InteropServices;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Identity.Web.Diagnostics
+{
+    internal class OsHelper
```
Comment (bgavrilMS) [inline src/Microsoft.Identity.Web.Diagnostics/OsHelper.cs:13]: Copied from MSAL.NET; Not sure this is the right place for it, but it seems to be the only "common" DLL.

Comment (jennyf19) [reply src/Microsoft.Identity.Web.Diagnostics/OsHelper.cs:13]: This was pulled from R9, which is now in .NET 8, but not available on older frameworks, so leaving it in place for now.

**Review by jennyf19 — APPROVED** [review]

**Review by jmprieur — APPROVED** [review]
LGTM
Thanks @bgavrilMS

**Pull request merged**